cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java

```diff
@@ -855,30 +857,56 @@ public NetworkPolicy generateNetworkPolicy() {
                 .withFrom(kafkaClusterPeer, entityOperatorPeer)
                 .build();
 
+        rules.add(replicationRule);
+
         // Free access to 9092, 9093 and 9094 ports
-        NetworkPolicyPort plainPort = new NetworkPolicyPort();
-        plainPort.setPort(new IntOrString(CLIENT_PORT));
+        if (listeners != null && listeners.getPlain() != null) {
+            NetworkPolicyPort plainPort = new NetworkPolicyPort();
+            plainPort.setPort(new IntOrString(CLIENT_PORT));
 
-        NetworkPolicyIngressRule plainRule = new NetworkPolicyIngressRuleBuilder()
-                .withPorts(plainPort)
-                .withFrom()
-                .build();
+            NetworkPolicyIngressRule plainRule = new NetworkPolicyIngressRuleBuilder()
+                    .withPorts(plainPort)
+                    .withFrom()
+                    .build();
+
+            rules.add(plainRule);
+        }
 
-        NetworkPolicyPort tlsPort = new NetworkPolicyPort();
-        tlsPort.setPort(new IntOrString(CLIENT_TLS_PORT));
+        if (listeners != null && listeners.getTls() != null) {
```

Comment:
Looks good. One suggestion - I would create one "big if" `if (listeners != null)` and put `and` conditions as nested.